File: frontend/src/lib/tags.ts
/**
 * Typed client for the tag surface.
 *
 * Tags are the user's own vocabulary over their movements — the one dimension
 * the bank does not supply. They are deliberately *not* categories: a movement
 * has exactly one category and any number of tags, which is why every total
 * grouped by tag overlaps and must never be drawn as parts of a whole.
 *
 * A transaction carries `tag_ids`, not embedded tag objects, so a name is
 * resolved against the tag list the page already has. `tagIndex` is the one
 * place that lookup is built.
 */

import { request } from "./api";

/** `investment` tags are the ones the investment metrics read; `plain` is the
 *  default and carries no behaviour beyond grouping. */
export type TagKind = "plain" | "investment";

export type Tag = {
    id: string;
    name: string;
    slug: string;
    color: string | null;
    kind: TagKind;
    created_at: string;
};

export type TagInput = { name: string; kind?: TagKind; color?: string | null };

/** 409 means the slug already exists. Worth distinguishing at the call site:
 *  "ya tienes una etiqueta con ese nombre" is a different sentence from "no
 *  pudimos guardar". `request` puts the status in the message. */
export function isDuplicateTagError(error: unknown): boolean {
    return error instanceof Error && error.message.startsWith("API 409");
}

export const tagsApi = {
    list: () => request<{ items: Tag[] }>("/api/tags").then((r) => r.items),

    create: (body: TagInput) =>
        request<Tag>("/api/tags", { method: "POST", body: JSON.stringify(body) }),

    update: (id: string, body: Partial<TagInput>) =>
        request<Tag>(`/api/tags/${id}`, { method: "PATCH", body: JSON.stringify(body) }),

    remove: (id: string) => request<unknown>(`/api/tags/${id}`, { method: "DELETE" }),

    /** Replaces the whole set on one transaction. Sending the full list rather
     *  than a diff means two sheets open at once cannot interleave into a state
     *  neither user asked for — the last save wins, visibly. */
    setForTransaction: (transactionId: string, tagIds: string[]) =>
        request<{ tag_ids: string[] }>(`/api/transactions/${transactionId}/tags`, {
            method: "PUT",
            body: JSON.stringify({ tag_ids: tagIds }),
        }),

    /** Adds one tag to many transactions. Additive on purpose: bulk tagging is
     *  "also mark these", never "make these the only tags". */
    attachTransactions: (tagId: string, transactionIds: string[]) =>
        request<unknown>(`/api/tags/${tagId}/transactions`, {
            method: "POST",
            body: JSON.stringify({ transaction_ids: transactionIds }),
        }),
};

export const { list: listTags, create: createTag } = tagsApi;

/** id -> tag, for turning a transaction's `tag_ids` into chips. */
export function tagIndex(tags: Tag[]): Map<string, Tag> {
    return new Map(tags.map((t) => [t.id, t]));
}

/** Case- and accent-insensitive contains, so "cafe" finds "Café". */
export function matchesTag(tag: Tag, query: string): boolean {
    return fold(tag.name).includes(fold(query));
}

/** True when no existing tag has exactly this name — i.e. "crear" is offerable. */
export function isNewTagName(tags: Tag[], query: string): boolean {
    const q = fold(query.trim());
    return q.length > 0 && !tags.some((t) => fold(t.name) === q);
}

function fold(value: string): string {
    return value
        .normalize("NFD")
        .replace(/[\u0300-\u036f]/g, "")
        .toLowerCase()
        .trim();
}

File: backend/src/tomin/adapters/inbound/http/blueprints/tags.py
"""Tag CRUD plus the two tagging gestures (docs/redesign-plan.md §7, B6).

``PUT /api/transactions/{id}/tags`` replaces one transaction's list; ``POST
/api/tags/{id}/transactions`` adds one tag to many transactions. Replace and
add-to-many are different operations on purpose -- bulk tagging must not strip
tags the selected rows already carry.
"""

from __future__ import annotations

from uuid import UUID

from flask import Blueprint, jsonify, request

from .....application.ports.outbound import DuplicateTagError
from .....application.use_cases.tags import UNSET
from ..auth import current_user_id, get_container
from ..serialization import tag_json

tags_bp = Blueprint("tags", __name__, url_prefix="/api/tags")

#: Mirrors the transaction patch: only declared fields are writable, and an
#: unrecognised key is a 400 rather than a silent no-op.
_PATCHABLE = {"name", "color", "kind"}


@tags_bp.get("")
def list_tags():
    user_id = current_user_id()
    items = get_container().manage_tags.list(user_id=user_id)
    return jsonify(items=[tag_json(t) for t in items], total=len(items))


@tags_bp.post("")
def create_tag():
    user_id = current_user_id()
    body = request.get_json(silent=True) or {}
    if not body.get("name"):
        return jsonify(error="name is required"), 400
    try:
        tag = get_container().manage_tags.create(
            user_id=user_id,
            name=body["name"],
            color=body.get("color"),
            kind=body.get("kind", "plain"),
        )
    except DuplicateTagError as err:
        # 409 rather than the ValueError handler's 400: the request is
        # well-formed, it collides with state that already exists.
        return jsonify(error="Tag already exists", detail=str(err)), 409
    return jsonify(tag_json(tag)), 201


@tags_bp.patch("/<tag_id>")
def update_tag(tag_id: str):
    user_id = current_user_id()
    body = request.get_json(silent=True) or {}
    unknown = set(body) - _PATCHABLE
    if unknown:
        return jsonify(error=f"Unsupported fields: {sorted(unknown)}"), 400
    try:
        tag = get_container().manage_tags.update(
            user_id=user_id,
            tag_id=UUID(tag_id),
            name=body.get("name", UNSET),
            color=body.get("color", UNSET),
            kind=body.get("kind", UNSET),
        )
    except DuplicateTagError as err:
        return jsonify(error="Tag already exists", detail=str(err)), 409
    return jsonify(tag_json(tag))


@tags_bp.delete("/<tag_id>")
def delete_tag(tag_id: str):
    user_id = current_user_id()
    untagged = get_container().manage_tags.delete(user_id=user_id, tag_id=UUID(tag_id))
    # The transactions survive; only the annotation is gone.
    return jsonify(tag_id=tag_id, transactions_untagged=untagged)


@tags_bp.post("/<tag_id>/transactions")
def tag_transactions(tag_id: str):
    """Attach one tag to many transactions. Additive, never a replacement."""
    user_id = current_user_id()
    body = request.get_json(silent=True) or {}
    ids = _uuid_list(body, "transaction_ids")
    tagged = get_container().manage_tags.attach_to_transactions(
        user_id=user_id, tag_id=UUID(tag_id), transaction_ids=ids
    )
    return jsonify(tag_id=tag_id, transactions_tagged=tagged)


def _uuid_list(body: dict, key: str) -> list[UUID]:
    raw = body.get(key)
    if not isinstance(raw, list):
        raise ValueError(f"'{key}' must be a list of ids")
    return [UUID(value) for value in raw]
